src/System.Management.Automation/engine/MshObjectIConvertible.cs

```diff
@@ -0,0 +1,104 @@
+        /// Otherwise, the result is the type code of the base object,
+        /// as determined by the base object's implementation of <see cref="System.IConvertible"/>.
+        /// </remarks>
+        /// <returns>Returns <see cref="System.TypeCode"/> of the PSObject.</returns>
+        public TypeCode GetTypeCode()
+        {
+            object obj = PSObject.Base(this);
+            return obj is PSObject ? TypeCode.Object : Convert.GetTypeCode(obj);
+        }
+
+        /// <inheritdoc/>
+        public bool ToBoolean(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToBoolean(provider);
+
+        /// <inheritdoc/>
+        public char ToChar(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToChar(provider);
+
+        /// <inheritdoc/>
+        public sbyte ToSByte(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToSByte(provider);
+
+        /// <inheritdoc/>
+        public byte ToByte(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToByte(provider);
+
+        /// <inheritdoc/>
+        public short ToInt16(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToInt16(provider);
+
+        /// <inheritdoc/>
+        public ushort ToUInt16(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToUInt16(provider);
+
+        /// <inheritdoc/>
+        public int ToInt32(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToInt32(provider);
+
+        /// <inheritdoc/>
+        public uint ToUInt32(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToUInt32(provider);
+
+        /// <inheritdoc/>
+        public long ToInt64(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToInt64(provider);
+
+        /// <inheritdoc/>
+        public ulong ToUInt64(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToUInt64(provider);
+
+        /// <inheritdoc/>
+        public float ToSingle(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToSingle(provider);
+
+        /// <inheritdoc/>
+        public double ToDouble(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToDouble(provider);
+
+        /// <inheritdoc/>
+        public decimal ToDecimal(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToDecimal(provider);
+
+        /// <inheritdoc/>
+        public DateTime ToDateTime(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToDateTime(provider);
+
+        /// <inheritdoc/>
+        public string ToString(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToString(provider);
+
+        /// <inheritdoc/>
+        public object ToType(Type conversionType, IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToType(conversionType, provider);
+
+        private static IConvertible GetIConvertibleOrThrow(PSObject pso)
+        {
+            object obj = PSObject.Base(pso);
+            if (obj is PSObject || obj is not IConvertible value)
+            {
+                ThrowInvalidCastException();
```

Comment:
Since this method is only used once here, would it be better to simply have this block throw the exception and not have the workaround?

Reply:
`throw` blocks inlining. .Net suggests long term patterns like ArgumentNullException.ThrowIf(). We can benefit from this too.